### PR TITLE
Implement more weapon stuff

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/logic/object/weapon.h [hunk 1]
```diff
@@ -158,11 +158,98 @@ class WeaponTemplate : public MemoryPoolObject
...
+        WeaponSlotType wslot,
+        int barrel,
+        const Object *victim_obj,
+        const Coord3D *victim_pos,
+        const WeaponBonus &bonus,
+        bool is_projectile_detonation,
+        bool ignore_ranges,
+        Weapon *firing_weapon,
+        ObjectID projectile_id,
+        bool dodamage) const;
```
Comment (xezon) [inline src/game/logic/object/weapon.h]: do_damage?

File: src/game/logic/object/weapon.h [hunk 2]
```diff
@@ -262,11 +351,125 @@ class Weapon : public MemoryPoolObject, public SnapShot
 
     Weapon &operator=(const Weapon &that);
 
-    bool Is_Within_Attack_Range(const Object *source, const Object *target) const;
-    bool Is_Within_Attack_Range(const Object *source, const Coord3D *target) const;
-    float Get_Attack_Range(const Object *source) const;
+    bool Is_Within_Attack_Range(const Object *source_obj, const Object *victim_obj) const;
+    bool Is_Within_Attack_Range(const Object *source_obj, const Coord3D *victim_obj) const;
```
Comment (xezon) [inline src/game/logic/object/weapon.h]: `victim_pos` or `target_pos`?

File: src/game/logic/object/weapon.h [hunk 3]
```diff
@@ -262,11 +351,125 @@ class Weapon : public MemoryPoolObject, public SnapShot
 
     Weapon &operator=(const Weapon &that);
 
-    bool Is_Within_Attack_Range(const Object *source, const Object *target) const;
-    bool Is_Within_Attack_Range(const Object *source, const Coord3D *target) const;
-    float Get_Attack_Range(const Object *source) const;
+    bool Is_Within_Attack_Range(const Object *source_obj, const Object *victim_obj) const;
```
Comment (xezon) [inline src/game/logic/object/weapon.h]: Maybe use `target_obj` like in functions below? Several times in this class for victim_obj, victim_pos.

File: src/game/logic/object/weapon.h [hunk 4]
```diff
@@ -262,11 +351,125 @@ class Weapon : public MemoryPoolObject, public SnapShot
...
+    bool Is_Within_Attack_Range(const Object *source_obj, const Coord3D *victim_obj) const;
+    bool Is_Within_Target_Pitch(const Object *source_obj, const Object *victim_obj) const;
+    bool Is_Damage_Weapon() const;
+    bool Is_Too_Close(const Object *source_obj, const Object *target_obj) const;
+    bool Is_Too_Close(const Object *source_obj, const Coord3D *target_pos) const;
+    bool Is_Clear_Firing_Line_Of_Sight_Terrain(const Object *source_obj, const Object *target_obj) const;
+    bool Is_Clear_Firing_Line_Of_Sight_Terrain(const Object *source_obj, const Coord3D &target_pos) const;
```
Comment (xezon) [inline src/game/logic/object/weapon.h:361]: In same rare functions Coord3D is passed by reference, and in other passed by pointer. Is this correct?

File: src/game/logic/object/weapon.h [hunk 5]
```diff
@@ -262,11 +351,125 @@ class Weapon : public MemoryPoolObject, public SnapShot
...
+    float Get_Attack_Distance(const Object *source_obj, const Object *victim_obj, const Coord3D *victim_pos) const;
+    float Get_Primary_Damage_Radius(const Object *source_obj) const;
+    int Get_Pre_Attack_Delay(const Object *source_obj, const Object *target_obj) const;
+    void Get_Firing_Line_Of_Sight_Origin(const Object *source_obj, Coord3D &pos) const;
+
+    void On_Weapon_Bonus_Change(const Object *source_obj);
+    void Load_Ammo_Now(const Object *source_obj);
+    void Reload_Ammo(const Object *source_obj);
+    void Compute_Bonus(const Object *source_obj, int flags, WeaponBonus &bonus) const;
+    void Set_Clip_Percent_Full(float percent, bool b);
```
Comment (xezon) [inline src/game/logic/object/weapon.h:388]: bool full maybe? Just guessing by the function name.

File: src/game/logic/object/weaponset.h [hunk 6]
```diff
@@ -142,23 +142,26 @@ class WeaponSet : public SnapShot
     virtual void Xfer_Snapshot(Xfer *xfer) override;
     virtual void Load_Post_Process() override {}
 
-    void Update_Weapon_Set(const Object *obj);
+    void Update_Weapon_Set(const Object *sourceobj);
```
Comment (xezon) [inline src/game/logic/object/weaponset.h]: `source_obj`? Many times in this class.

File: src/game/logic/object/weaponset.h [hunk 7]
```diff
@@ -142,23 +142,26 @@ class WeaponSet : public SnapShot
...
+    CanAttackResult Get_Able_To_Attack_Specific_Object(AbleToAttackType type,
+        const Object *sourceobj,
+        const Object *victimobj,
+        CommandSourceType source,
+        WeaponSlotType slot) const;
     CanAttackResult Get_Able_To_Use_Weapon_Against_Target(AbleToAttackType type,
-        const Object *obj,
-        const Object *obj2,
+        const Object *sourceobj,
+        const Object *victimobj,
```
Comment (xezon) [inline src/game/logic/object/weaponset.h]: `victim_obj`? Many times in this class.

File: src/game/logic/object/weaponset.cpp [hunk 8]
```diff
@@ -214,50 +215,217 @@ bool WeaponSet::Has_Any_Damage_Weapon() const
...
+        if (lock == LOCKED_LEVEL_2) {
+            m_curWeaponLocked = NOT_LOCKED;
+        } else if (lock == LOCKED_LEVEL_1) {
+            if (m_curWeaponLocked == LOCKED_LEVEL_1) {
+                m_curWeaponLocked = NOT_LOCKED;
+            }
+        } else {
+            captainslog_dbgassert(false, "calling releaseWeaponLock with NOT_LOCKED makes no sense. why did you do this?");
+        }
+    }
```
Comment (xezon) [inline src/game/logic/object/weaponset.cpp:374]: No else assert here?

File: src/game/logic/object/weapon.h [hunk 9]
```diff
@@ -262,11 +351,125 @@ class Weapon : public MemoryPoolObject, public SnapShot
...
+
+    void Set_Leech_Range_Active(bool set) { m_leechWeaponRangeActive = set; }
+    void Set_Next_Shot(unsigned int time) { m_whenWeCanFireAgain = time; }
+    void Set_Status(WeaponStatus status) { m_status = status; }
+    void Set_Pre_Attack_Finished(bool set) { m_whenPreAttackFinished = set; }
+    void Set_Max_Shot_Count(int count) { m_maxShotCount = count; }
+
+    bool Has_Laser() const { return m_template->Get_Laser_Name().Is_Not_Empty(); }
+
+    float Estimate_Weapon_Damage(const Object *source_obj, const Object *victim_obj)
```
Comment (xezon) [inline src/game/logic/object/weapon.h:469]: Perhaps remove this function and just declare above function as
```
float Estimate_Weapon_Damage(const Object *source_obj, const Object *victim_obj, const Coord3D *victim_pos = nullptr);
```

Notice the `const Coord3D *victim_pos = nullptr`

File: src/game/logic/object/weapon.h (hunk 1 above)
Comment (jonwil) [reply src/game/logic/object/weapon.h]: fixed.

File: src/game/logic/object/weapon.h (hunk 2 above)
Comment (jonwil) [reply src/game/logic/object/weapon.h]: fixed.

File: src/game/logic/object/weapon.h (hunk 3 above)
Comment (jonwil) [reply src/game/logic/object/weapon.h]: fixed.

Comment (jonwil) [reply src/game/logic/object/weapon.h]: Evaluated and looked against the original and used target/victim as appropriate.

File: src/game/logic/object/weapon.h (hunk 4 above)
Comment (jonwil) [reply src/game/logic/object/weapon.h:361]: Yes this is correct.

File: src/game/logic/object/weaponset.h (hunk 6 above)
Comment (jonwil) [reply src/game/logic/object/weaponset.h]: fixed.

File: src/game/logic/object/weapon.h (hunk 5 above)
Comment (jonwil) [reply src/game/logic/object/weapon.h:388]: Nope, its not that.

File: src/game/logic/object/weaponset.h (hunk 7 above)
Comment (jonwil) [reply src/game/logic/object/weaponset.h]: Fixed.

File: src/game/logic/object/weaponset.cpp (hunk 8 above)
Comment (jonwil) [reply src/game/logic/object/weaponset.cpp:374]: none in the original.

File: src/game/logic/object/weapon.h (hunk 9 above)
Comment (jonwil) [reply src/game/logic/object/weapon.h:469]: This exists in the original and I intend to keep it that way.

**Review by xezon — APPROVED** [review]
I briefly tested in Debug on Map with all faction units shooting onto practice targets. Worked without crashes or obvious issues. We can do more detailed test on last Weapon code iteration.

**Pull request merged**